video/out/vo_opengl_cb.c

```diff
@@ -475,18 +508,30 @@ static int control(struct vo *vo, uint32_t request, void *data)
     case VOCTRL_SET_PANSCAN:
         pthread_mutex_lock(&p->ctx->lock);
         p->ctx->force_update = true;
-        update(p);
+        update(p->ctx);
         pthread_mutex_unlock(&p->ctx->lock);
         return VO_TRUE;
     case VOCTRL_UPDATE_RENDER_OPTS:
         pthread_mutex_lock(&p->ctx->lock);
         p->ctx->update_new_opts = true;
-        update(p);
+        update(p->ctx);
         pthread_mutex_unlock(&p->ctx->lock);
         return VO_TRUE;
+    case VOCTRL_PERFORMANCE_DATA:
+        gl_video_perfdata(p->ctx->renderer, (struct voctrl_performance_data *)data);
```

Comment:
I don't think this can actually work. Wouldn't you access the GL context from an unrelated thread?

Reply:
you are right that this shouldn't work, looking at the comment in the struct at the top. it does work though besides [this](https://github.com/mpv-player/mpv/issues/4615#issuecomment-326738862). it would be a big pity if we can't that data with cocoa-cb.

Reply:
Well, it's just not sane. It looks like the gl_video call does not actually do any GL calls, which is why this appears to work. But it will still do unsynchronized accesses from 2 threads (the VO thread and the renderer thread). There are two possibilities how to fix this:
1. retrieve the stats after each renderer call, put them into a mutex protected variable, read the copy with VOCTRL (doesn't work, see below)
2. put the performance data into an explicitly shared struct which is protected by its own mutex (and maybe refcounted), which would mean rewriting half the perf reporting code path

Also major WTF: voctrl_performance_data points to allocated strings which the GL code can randomly during rendering! So using this is just a guarantee for undefined behavior (even for normal vp_gpu on other OSes). This mixes undefined behavior with race conditions. This is just insane. I have no idea how I ever (apparently) agreed to such garbage.

So yeah, don't use this, it'll just crash you in obscure situations. If you implement it, it will not only be wrong, but the next person who fixes above bug will have to rewrite your new code too.

Reply:
i had hoped to make stats.lua completely working since it's an embedded feature now. though not much i can do about it.

should a open an issue about this?

Reply:
Yeah. Something about how the VOCTRL is not thread safe. I'm not actually sure how much we even care about this, since apparently the timers are inaccurate anyway.